File: PA4/List.c
/*
Rajdipa Chowdhury
cruzid - rachowdh
pa 4
*/
#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj{
  int data;
  struct NodeObj* next;
  struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj {
  Node front;
  Node back;
  Node cursor;
  int index;
  int length;
} ListObj;

/*****CONSTRUCTORS - DESTRUCTORS*****/
// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   return(N);
}
// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}
List newList(void){
  List L;
  L = malloc(sizeof(ListObj));
  L->front = L->back = L->cursor = NULL;
  L->length = 0;
  L->index = -1;
  return(L);
}
// freeQueue()
// Frees all heap memory associated with Queue *pQ, and sets *pQ to NULL.S
void freeList(List* pL){
  if(pL != NULL && *pL != NULL){
    Node cur = (*pL)->front;
    while(cur != NULL){
      deleteFront(*pL);
      cur = (*pL)->front;
    }
    free(cur);
  }
  free(*pL);
  *pL = NULL;
}

// Access functions ---
int length(List L){
  if(L == NULL){
    printf("Invalid: calling length() on NULL List Reference");
    exit(1);
  }
  return(L->length);
}

int index(List L){
  if(L == NULL){
    printf("Invalid: calling index() on NULL List Reference");
    exit(1);
  }
  return(L->index);
}

int front(List L){
  if(L == NULL){
    printf("Invalid: calling front() on NULL List Reference");
    exit(1);
  }
  return(L->front->data);
}

int back(List L){
  if(L == NULL){
    printf("Invalid: calling back() on NULL List Reference");
    exit(1);
  }
  return(L->back->data);
}

int get(List L)
{
   if(L == NULL)
    {
      printf("Invalid: calling get() on NULL List reference");
      exit(1);
   }
   if(L->index < 0)
   {
      printf("Invalid: calling get() with an undefined index on List");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("Invalid: calling get() on empty List reference");
      exit(1);
   }
   return L->cursor->data;
}

int equals(List A, List B){
  if(A == NULL || B == NULL){
    printf("Invalid: Calling equals() on NULL List Reference");
    exit(1);
  }

  if(length(A) != length(B)){
    return(0);
  }
  Node an = A->front;
  Node bn = B->front;

  while(an != NULL){
    if(an->data != bn->data){
      an = bn = NULL;
      return(0);
    }
    else {
      an = an->next;
      bn = bn->next;
    }
  }
  return(1);
}

// Manipulation procedures ----------------------------------------------------

void clear(List L){
  if(L == NULL){
    printf("Invalid: Caling clear() on NULL List Reference");
    exit(1);
  }
  Node cur = L->front;
  while(cur != NULL){
    deleteFront(L);
    cur = L->front;
  }
  freeNode(&cur);
  L->front = L->back = L->cursor = NULL;
  L->length = 0;
  L->index = -1;
}

void moveFront(List L){
  if(L == NULL){
    printf("Invalid: Calling moveFront() on NULL List reference");
    exit(1);
  }
  if(L->length != 0){
    L->cursor = L->front;
    L->index = 0;
  }
}

void moveBack(List L){
  if(L == NULL){
    printf("Invalid: Calling moveBack() on NULL List reference");
    exit(1);
  }
  if(L->length != 0){
    L->cursor = L->back;
    L->index = L->length - 1;
  }
}

void movePrev(List L){
  if(L == NULL){
    printf("Invalid: Calling movePrev() on NULL List reference");
    exit(1);
  }
  if(L->cursor != NULL && L->index != 0){
    L->cursor = L->cursor->prev;
    L->index--;
  }else if(L->cursor != NULL && L->index == 0){
    L->cursor = NULL;
    L->index = -1;
  }
}

void moveNext(List L){
  if(L == NULL){
    printf("Invalid: Calling moveNext() on NULL List reference");
    exit(1);
  }
  if(L->cursor != NULL && L->index != L->length - 1){
    L->cursor = L->cursor->next;
    L->index++;
  }else if(L->cursor != NULL && L->index == L->length - 1){
    L->index = -1;
    L->cursor = NULL ;
  }
}
void prepend(List L,int data){
  if(L == NULL){
   printf("Invalid: Calling prepend() on NULL List reference");
   exit(1);
  }
  Node temp = newNode(data);
  temp->prev = NULL;
  temp->next = L->front;
  if(L->length > 0)
    L->front->prev = temp;
  L->front = temp;
  L->length++;
  if(L->cursor != NULL)
    L->index++;
  if(L->length == 1)
    L->back = temp;
}

void append(List L, int data){
  if(L == NULL){
    printf("Invalid: Calling append() on NULL List reference");
    exit(1);
  }
  Node temp = newNode(data);

  if(L->length == 0){
    L->front = temp;
    L->back = temp;
    L->length++;
  }else{
    L->back->next = temp;
    temp->prev = L->back;
    L->back = temp;
    L->length++;
  }
}

void insertBefore(List L, int data){
  if(L == NULL ){
    printf("Invalid: Calling insertBefore() on NULL List reference");
    exit(1);
  }
  if (L->cursor == NULL){
    printf("Invalid: Calling insertBefore() on undefined cursor");
    exit(1);
  }
  if(L->index == 0)
    prepend(L,data);
  else {
    Node temp = newNode(data);
    temp->prev = L->cursor->prev;
    temp->next = L->cursor;
    L->cursor->prev->next = temp;
    L->cursor->prev = temp;
    L->index++;
    L->length++;
  }
}

void insertAfter(List L, int data){
  if(L == NULL ){
    printf("Invalid: Calling insertAfter() on NULL List reference");
    exit(1);
  }
  if (L->cursor == NULL){
    printf("Invalid: Calling insertAfter() on undefined cursor");
    exit(1);
  }
  if(L->index == L->length - 1){
    append(L,data);
  }else {
    Node temp = newNode(data);
    L->cursor->next->prev = temp;
    temp->next = L->cursor->next;
    L->cursor->next = temp;
    temp->prev = L->cursor;
    L->length++;
  }
}

void deleteFront(List L){

  if(L == NULL){
    printf("Invalid: Calling deleteFront() on NULL List reference");
    exit(1);
  }
  if(L->length == 0){
    printf("Invalid: Calling deleteFront() on empty List");
    exit(1);
  }
  Node temp = L->front;

  if(L->length > 1){
    L->front = L->front->next;
  } else{
    L->front = L->back = NULL;
  }
  freeNode(&temp);
  L->length--;
  if(L->index == 0){
    L->index = -1;
    L->cursor = NULL;
  }
  if(L->cursor != NULL)
  {
    L->index--;
  }
}

void deleteBack(List L){
  if(L == NULL){
    printf("Invalid: Calling deleteBack() on NULL List reference");
    exit(1);
  }else if(L->length == 0){
    printf("Invalid: calling deleteBack() on an empty list");
    exit(1);
  }else if(L->length == 1){
    if(L->index == 0){
      L->index = -1;
      L->cursor = NULL;
    }
    freeNode(&L->front);
    L->front = NULL;
    L->back = NULL;
    L->length = 0;
  }else{
    if(L->cursor == L->back){
      L->cursor = NULL;
      L->index = -1;
    }
    Node temp = L->back;
    L->back = L->back->prev;
    L->back->next = NULL;
    freeNode(&temp);
    L->length--;
  }
}
void delete(List L){
  if(L == NULL ){
    printf("Invalid: Calling delete() on NULL List referece");
    exit(1);
  }
  if( L->cursor == NULL){
    printf("Invalid: Calling delete() on undefined cursor");
    exit(1);
  }

  if(L->index == 0)
    deleteFront(L);
  else{
    Node oNd = L->cursor;
    oNd->prev->next = oNd->next;
    if(L->index == L->length - 1)
      L->back = oNd->prev;
    else
      oNd->next->prev = oNd->prev;
    freeNode(&oNd);
    L->cursor = NULL;
    L->length--;
    L->index = -1;
  }
}
// Other operations
void printList(FILE *out, List L){
  moveFront(L);
  while(L->cursor != NULL){
    fprintf(out, "%d ",get(L));
    moveNext(L);
  }
}

List copyList(List L){
  if(L == NULL){
    printf("Invalid: Calling copyList() on Null reference");
    exit(1);
  }
  List tL = newList();
  Node cur = L->front;
  while(cur != NULL){
    append(tL,cur->data);
    cur = cur->next;
  }
  return(tL);
}

File: PA1/List.java
/*
Rajdipa Chowdhury
cruz-iD - rachowdh
PA-1
*/
public class List{
  private class Node{
    int data;
    Node prev;
    Node next;

    //Constructor
      Node(int data_n, Node next_n , Node prev_n){
        this.data = data_n;
        this.next = next_n;
        this.prev = prev_n;
      }
    // toString(): Overrides Object's toString()method
      public String toString(){
        return String.valueOf(data);
      }
  }
private Node front;
private Node back;
private Node cursor;
private int length;
private int index;

// Creates a new empty list
List(){
  cursor = null;
  back = null;
  front = null;
  length = 0 ;
  index = -1;
}
// Access Functions
// Returns the number of elements in this List
int length(){
   return length;
}
// If cursor is defined, returns the index of the cursor element,
 // otherwise returns -1.
 int index(){
   if(cursor == null){
     return -1 ;
   }
   else{
     return index;
   }
 }

 // Returns front element. Pre: length()>0
 int front(){
   if (length() <= 0 ){
     throw new RuntimeException(
              "List Error: front() called on empty List");
   }
   return front.data;
 }
// Returns back element. Pre: length()>0
 int back(){
   if (length() <= 0 ){
     throw new RuntimeException(
              "List Error: back() called on empty List");
   }
   return back.data;

 }
 // Returns cursor element. Pre: length()>0, index()>=0
 int get(){
   if(length() <= 0){
     throw new RuntimeException(
            "List Error: get called on empty List");
   }
   if (index()<0){
      throw new RuntimeException(
         "List Error: get() called on undefined cursor");
       }
       return cursor.data;

 }

 // Returns true if and only if this List and L are the same
 // integer sequence. The states of the cursors in the two Lists
 // are not used in determining equality.
 boolean equals(List L){
   Node temp1 = front;
   Node temp2 = L.front;
   boolean istrue = true;
   if (L.length() != length ){
     return false;
   }

   else{
            while(temp1 != null && temp2 != null){
              if(temp1.data != temp2.data ) {
                return false;
                }
                else{
                  temp1 = temp1.next;
                  temp2 = temp2.next;
                }
            }
            return istrue;

          }
        }
    // Manipulation procedures
    // Resets this List to its original empty state
    void clear(){
      cursor = null;
      front = null;
      back = null;
      length = 0;
      index = -1;
    }
    // If List is non-empty, places the cursor under the front element,
 // otherwise does nothing.
 void moveFront(){
   if(length > 0){
     cursor = front;
     index = 0;
   }
 }

 // If List is non-empty, places the cursor under the back element,
 // otherwise does nothing.
 void moveBack(){
   if(length > 0){
     cursor = back;
     index = length - 1 ;
   }
 }

 // If cursor is defined and not at front, moves cursor one step toward
 // front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing
void movePrev(){
  // If cursor is defined and not at front
  if (cursor != null && index != 0){
    cursor = cursor.prev;
    index--;
  }
  //if cursor is defined and at front
  else if (cursor != null && index == 0){
    cursor = null;
    index = -1;
  }
}

// If cursor is defined and not at back, moves cursor one step toward
 // back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing

void moveNext(){
  // If cursor is defined and not at back
  if (cursor != null && index != length -1 ){
    cursor = cursor.next;
    index++;
  }
  else if(cursor != null && index == length -1 ){
    cursor = null;
    index = -1;
  }

}

// Insert new element into this List. If List is non-empty,
 // insertion takes place before front element
 void prepend(int data){
   Node temp = new Node(data, front, null);
   if (front == null){
     back = temp;
   }
   else {
     front.prev = temp;
   }
   front = temp;
   length++ ;
 }

 // Insert new element into this List. If List is non-empty,
 // insertion takes place after back element

 void append(int data){
   Node temp = new Node(data,null,back);
   if (front == null){
     front = temp;
   }
   else{
     back.next = temp;
   }
   back = temp;
   length++ ;
 }

 // Insert new element before cursor.
// Pre: length()>0, index()>=0
void insertBefore(int data){
  Node temp = new Node(data,cursor,cursor.prev);
  if (length< 1 ){
    throw new RuntimeException(
              "list error:Cannot call inserBefore() on an empty list");
  }
  if ( cursor == null){
    throw new RuntimeException(
        "list error:Cannot call insertBefore() on an undefined cursor");
  }
  if (cursor.prev != null){
    cursor.prev.next = temp;
  }
  else{
    front = temp;
  }
  cursor.prev = temp;
  length ++;
}

// Inserts new element after cursor.
// Pre: length()>0, index()>=0

void insertAfter(int data){
  Node temp = new Node(data,cursor.next,cursor);
  if (length< 1 ){
    throw new RuntimeException(
              "list error:Cannot call inserAfter() on an empty list");
  }
  if ( cursor == null){
    throw new RuntimeException(
        "list error:Cannot call insertAfter() on an undefined cursor");
  }
  if (cursor.next != null){
    cursor.next.prev = temp;

  }
  else{
    back = temp;
  }
  cursor.next = temp;
  length++ ;
  }
  // Deletes the front element. Pre: length()>0
  void deleteFront(){
    if (length < 1 ){
      throw new RuntimeException(
                "list error:Cannot call deleteFront() on an empty list");
    }
    if(cursor == front ){
      cursor = null;
      index = -1;
    }
    front = front.next;
    front.prev = null;
    length--;
    }

    // Deletes the back element. Pre: length()>0

    void deleteBack(){
      if (length < 1 ){
        throw new RuntimeException(
                  "list error:Cannot call deleteBack() on an empty list");
      }
      if (cursor == back){
        cursor = null;
        index = -1;
      }
      back = back.prev;
      back.next = null;
      length-- ;
      }
      // Deletes cursor element, making cursor undefined.
 	    // Pre: length()>0, index()>=0

      void delete(){
        if (length < 1 ){
          throw new RuntimeException(
                    "list error:Cannot call delete() on an empty list");
        }
        if (cursor == null ){
          throw new RuntimeException(
                    "list error:Cannot call delete() on undefined");
        }
        if (cursor == front){
          deleteFront();
        }
        else if (cursor == back){
          deleteBack();
        }
        else {
          cursor.prev.next = cursor.next;
          cursor.next.prev = cursor.prev;
          cursor = null;
          index = -1;
          length--;
        }
      }


  // Overrides Object's toString method. Returns a String
 // representation of this List consisting of a space
// separated sequence of integers, with front on left

 public String toString(){
   Node tempfront = front;
   String strnew = new String();
   while(tempfront != null){
     strnew = strnew+ String.valueOf(tempfront.data) + " " ;
     tempfront = tempfront.next;
   }
 return strnew;

 }
 // Returns a new List representing the same integer sequence as this
  // List. The cursor in the new list is undefined, regardless of the
  // state of the cursor in this List. This List is unchanged.
 List copy(){
   List newlist = new List();
   Node tempfront = front;
   while(tempfront != null){
     newlist.append(tempfront.data);
     tempfront = tempfront.next;
   }
   newlist.cursor = null;
   newlist.index = -1;
   return newlist;
 }








}
